com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl

```diff
@@ -292,40 +314,99 @@ inline void InitializeBRDFData(half3 albedo, half metallic, half3 specular, half
-
-    outBRDFData.normalizationTerm = outBRDFData.roughness * 4.0h + 2.0h;
-    outBRDFData.roughness2MinusOne = outBRDFData.roughness2 - 1.0h;
+    outBRDFData.roughness           = max(PerceptualRoughnessToRoughness(outBRDFData.perceptualRoughness), HALF_MIN_SQRT);
+    outBRDFData.roughness2          = max(outBRDFData.roughness * outBRDFData.roughness, HALF_MIN);
+    outBRDFData.normalizationTerm   = outBRDFData.roughness * 4.0h + 2.0h;
+    outBRDFData.roughness2MinusOne  = outBRDFData.roughness2 - 1.0h;
 
 #ifdef _ALPHAPREMULTIPLY_ON
     outBRDFData.diffuse *= alpha;
     alpha = alpha * oneMinusReflectivity + reflectivity;
 #endif
 }
 
+// Approximation of ConvertF0ForAirInterfaceToF0ForClearCoat15
+half3 ConvertF0ForAirInterfaceToF0ForClearCoat15Fast(half3 f0)
+{
+    // Approximation of iorTof0(f0ToIor(f0), 1.5)
+    // This assumes that the clear coat layer has an IOR of 1.5
+    return saturate(f0 * (f0 * 0.526868h + 0.529324h) - 0.0482256h);
+}
+
+half3 ConvertF0ForClearCoat15(half3 f0)
+{
+#if defined(SHADER_API_MOBILE)
+    return ConvertF0ForAirInterfaceToF0ForClearCoat15Fast(f0);
+#else
+    return ConvertF0ForAirInterfaceToF0ForClearCoat15(f0);
+#endif
+}
+
+inline void InitializeBRDFDataClearCoat(half clearCoatStrength, half clearCoatSmoothness, inout BRDFData baseBRDFData, out BRDFData outBRDFData)
+{
+    // Calculate Roughness of Clear Coat layer
+    outBRDFData.diffuse             = kDielectricSpec.aaa; // 1 - kDielectricSpec
+    outBRDFData.specular            = kDielectricSpec.rgb;
+
+    outBRDFData.perceptualRoughness = PerceptualSmoothnessToPerceptualRoughness(clearCoatSmoothness);
+    outBRDFData.roughness           = max(PerceptualRoughnessToRoughness(outBRDFData.perceptualRoughness), HALF_MIN_SQRT);
+    outBRDFData.roughness2          = max(outBRDFData.roughness * outBRDFData.roughness, HALF_MIN);
+    outBRDFData.normalizationTerm   = outBRDFData.roughness * 4.0h + 2.0h;
+    outBRDFData.roughness2MinusOne  = outBRDFData.roughness2 - 1.0h;
+    outBRDFData.grazingTerm         = saturate(clearCoatSmoothness + kDielectricSpec.x);
+
+// Relatively small effect, cut it for lower quality
+#if !defined(SHADER_API_MOBILE)
+    // Modify Roughness of base layer using coat IOR
+    half ieta                        = lerp(1.0h, CLEAR_COAT_IETA, clearCoatStrength);
+    half coatRoughnessScale          = Sq(ieta);
+    half sigma                       = RoughnessToVariance(PerceptualRoughnessToRoughness(baseBRDFData.perceptualRoughness));
+
+    baseBRDFData.perceptualRoughness = RoughnessToPerceptualRoughness(VarianceToRoughness(sigma * coatRoughnessScale));
+
+    // Recompute base material for new roughness, previous computation should be eliminated by the compiler (as it's unused)
+    baseBRDFData.roughness          = max(PerceptualRoughnessToRoughness(baseBRDFData.perceptualRoughness), HALF_MIN);
+    baseBRDFData.roughness2         = max(baseBRDFData.roughness * baseBRDFData.roughness, HALF_MIN);
```

Comment:
Shouldn't this be HALF_MIN_SQRT?

Reply:
I wonder if we shouldn't add a helper function for this, or just modify the default function (I've asked in #dev-renderpipeline slack as well)
half PerceptualRoughnessToRoughnessClamped(half perceptualRoughness)
{
    /// 0.089 perceptual roughness is the min value we can represent in fp16
    // to avoid denorm/division by zero as we need to do 1 / (pow(perceptualRoughness, 4)) in GGX
    return PerceptualRoughnessToRoughness(max(perceptualRoughness, 0.089));
}

Reply:
Yes, it should be min SQRT. 
We could clamp the perceptual roughness directly once and assume the rest of the roughness params are computed in known way.
I would do a separate function, so it's known it's special.

Reply:
fixed